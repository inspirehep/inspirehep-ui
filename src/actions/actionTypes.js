export const SEARCH_REQUEST = 'SEARCH_REQUEST';
export const SEARCH_SUCCESS = 'SEARCH_SUCCESS';
export const SEARCH_ERROR = 'SEARCH_ERROR';

export const LITERATURE_REQUEST = 'LITERATURE_REQUEST';
export const LITERATURE_SUCCESS = 'LITERATURE_SUCCESS';
export const LITERATURE_ERROR = 'LITERATURE_ERROR';
export const LITERATURE_REFERENCES_REQUEST = 'LITERATURE_REFERENCES_REQUEST';
export const LITERATURE_REFERENCES_SUCCESS = 'LITERATURE_REFERENCES_SUCCESS';
export const LITERATURE_REFERENCES_ERROR = 'LITERATURE_REFERENCES_ERROR';
export const LITERATURE_AUTHORS_REQUEST = 'LITERATURE_AUTHORS_REQUEST';
export const LITERATURE_AUTHORS_SUCCESS = 'LITERATURE_AUTHORS_SUCCESS';
export const LITERATURE_AUTHORS_ERROR = 'LITERATURE_AUTHORS_ERROR';

export const EXCEPTIONS_REQUEST = 'EXCEPTIONS_REQUEST';
export const EXCEPTIONS_SUCCESS = 'EXCEPTIONS_SUCCESS';
export const EXCEPTIONS_ERROR = 'EXCEPTIONS_ERROR';

export const INSPECT_REQUEST = 'INSPECT_REQUEST';
export const INSPECT_SUCCESS = 'INSPECT_SUCCESS';
export const INSPECT_ERROR = 'INSPECT_ERROR';
