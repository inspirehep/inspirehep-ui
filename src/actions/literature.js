import {
  LITERATURE_ERROR,
  LITERATURE_REQUEST,
  LITERATURE_SUCCESS,
  LITERATURE_REFERENCES_ERROR,
  LITERATURE_REFERENCES_REQUEST,
  LITERATURE_REFERENCES_SUCCESS,
  LITERATURE_AUTHORS_ERROR,
  LITERATURE_AUTHORS_REQUEST,
  LITERATURE_AUTHORS_SUCCESS,
} from './actionTypes';

function fetchingLiterature(recordId) {
  return {
    type: LITERATURE_REQUEST,
    payload: { recordId },
  };
}

function fetchLiteratureSuccess(result) {
  return {
    type: LITERATURE_SUCCESS,
    payload: result,
  };
}

function fetchLiteratureError(error) {
  return {
    type: LITERATURE_ERROR,
    payload: error,
  };
}

function fetchingLiteratureReferences() {
  return {
    type: LITERATURE_REFERENCES_REQUEST,
  };
}

function fetchLiteratureReferencesSuccess(result) {
  return {
    type: LITERATURE_REFERENCES_SUCCESS,
    payload: result,
  };
}

function fetchLiteratureReferencesError(error) {
  return {
    type: LITERATURE_REFERENCES_ERROR,
    payload: error,
  };
}

function fetchingLiteratureAuthors() {
  return {
    type: LITERATURE_AUTHORS_REQUEST,
  };
}

function fetchLiteratureAuthorsSuccess(result) {
  return {
    type: LITERATURE_AUTHORS_SUCCESS,
    payload: result,
  };
}

function fetchLiteratureAuthorsError(error) {
  return {
    type: LITERATURE_AUTHORS_ERROR,
    payload: error,
  };
}

export function fetchLiterature(recordId) {
  return async (dispatch, getState, http) => {
    dispatch(fetchingLiterature(recordId));
    try {
      const response = await http.get(`/literature/${recordId}`, {
        headers: {
          Accept: 'application/vnd+inspire.record.ui+json',
        },
      });
      dispatch(fetchLiteratureSuccess(response.data));
    } catch (error) {
      dispatch(fetchLiteratureError(error.data));
    }
  };
}

export function fetchLiteratureReferences(recordId) {
  return async (dispatch, getState, http) => {
    dispatch(fetchingLiteratureReferences());
    try {
      const response = await http.get(`/literature/${recordId}/references`);
      dispatch(fetchLiteratureReferencesSuccess(response.data));
    } catch (error) {
      dispatch(fetchLiteratureReferencesError(error.data));
    }
  };
}

export function fetchLiteratureAuthors(recordId) {
  return async (dispatch, getState, http) => {
    dispatch(fetchingLiteratureAuthors());
    try {
      const response = await http.get(`/literature/${recordId}/authors`);
      dispatch(fetchLiteratureAuthorsSuccess(response.data));
    } catch (error) {
      dispatch(fetchLiteratureAuthorsError(error.data));
    }
  };
}
