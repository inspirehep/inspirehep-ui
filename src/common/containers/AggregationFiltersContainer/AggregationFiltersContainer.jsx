import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Immutable from 'immutable';

import AggregationFilter from '../../components/AggregationFilter';
import { forceArray } from '../../utils';
import search from '../../../actions/search';
import './AggregationFiltersContainer.scss';

const RANGE_AGGREATION_KEY = 'earliest_date';

class AggregationFiltersContainer extends Component {
  static isRange(aggregationKey) {
    return aggregationKey === RANGE_AGGREATION_KEY;
  }

  static compareAggregationEntries([, aggregation1], [, aggregation2]) {
    const order1 = aggregation1.getIn(['meta', 'order']);
    const order2 = aggregation2.getIn(['meta', 'order']);
    return order1 - order2;
  }

  onAggregationChange(key, selections) {
    const isRange = AggregationFiltersContainer.isRange(key);
    let aggregations = selections;
    if (isRange && aggregations.length > 0) {
      aggregations = selections.join('--');
    }
    this.props.onAggregationChange(key, aggregations);
  }

  render() {
    return (
      <div className="__AggregationFiltersContainer__ bg-white pa3">
        {this.props.aggregations
          .entrySeq()
          .filter(([, aggregation]) => aggregation.get('buckets').size > 0)
          .sort(AggregationFiltersContainer.compareAggregationEntries)
          .map(([aggregationKey, aggregation]) => (
            <div key={aggregationKey}>
              <AggregationFilter
                range={aggregationKey === RANGE_AGGREATION_KEY}
                name={aggregation.getIn(['meta', 'title'])}
                buckets={aggregation.get('buckets')}
                selections={forceArray(this.props.query[aggregationKey])}
                onChange={selections => {
                  this.onAggregationChange(aggregationKey, selections);
                }}
              />
            </div>
          ))}
      </div>
    );
  }
}

AggregationFiltersContainer.propTypes = {
  onAggregationChange: PropTypes.func.isRequired,
  aggregations: PropTypes.instanceOf(Immutable.Map).isRequired,
  query: PropTypes.objectOf(PropTypes.any).isRequired,
};

const stateToProps = state => ({
  aggregations: state.search.get('aggregations'),
  query: state.router.location.query,
});

export const dispatchToProps = dispatch => ({
  onAggregationChange(aggregationKey, selections) {
    dispatch(search({ [aggregationKey]: selections }));
  },
});

export default connect(stateToProps, dispatchToProps)(
  AggregationFiltersContainer
);
