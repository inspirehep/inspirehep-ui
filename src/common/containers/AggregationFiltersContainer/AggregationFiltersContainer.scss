.__AggregationFiltersContainer__ {
  background: #fff;
  padding: 10px;
}
