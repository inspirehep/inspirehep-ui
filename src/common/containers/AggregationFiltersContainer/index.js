import AggregationFiltersContainer  from './AggregationFiltersContainer';

export default AggregationFiltersContainer;
