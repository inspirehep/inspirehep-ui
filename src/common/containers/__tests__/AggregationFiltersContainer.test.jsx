import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import { getStore, getStoreWithState } from '../../../fixtures/store';
import { SEARCH_REQUEST } from '../../../actions/actionTypes';
import AggregationFiltersContainer, {
  dispatchToProps,
} from '../AggregationFiltersContainer/AggregationFiltersContainer';

describe('AggregationFiltersContainer', () => {
  it('renders initial state with initial url query q param', () => {
    const store = getStoreWithState({
      router: { location: { query: { agg1: 'agg1-selected' } } },
      search: fromJS({
        aggregations: {
          agg1: {
            buckets: [{}],
            meta: {
              title: 'Jessica Jones',
              order: 1,
            },
          },
          agg2: {
            buckets: [{}],
            meta: {
              title: 'Luke Cage',
              order: 2,
            },
          },
        },
      }),
    });
    const wrapper = shallow(
      <AggregationFiltersContainer store={store} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render aggregations with empty buckets', () => {
    const store = getStoreWithState({
      search: fromJS({
        aggregations: {
          agg: {
            buckets: [{}],
            meta: {
              title: 'Jessica Jones',
              order: 1,
            },
          },
          emptyAgg: {
            buckets: [],
            meta: {
              title: 'Luke Cage',
              order: 2,
            },
          },
        },
      }),
    });
    const wrapper = shallow(
      <AggregationFiltersContainer store={store} />
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  // TODO: test onAggregationChange when range aggregation

  it('dispatches search onAggregationChange', () => {
    const store = getStore();
    const props = dispatchToProps(store.dispatch);
    props.onAggregationChange('agg1', ['selected']);
    const actions = store.getActions();
    const expectedAction = actions.find(
      action => action.type === SEARCH_REQUEST
    );
    expect(expectedAction).toBeDefined();
    expect(expectedAction.payload).toEqual({ agg1: ['selected'] });
  });
});
