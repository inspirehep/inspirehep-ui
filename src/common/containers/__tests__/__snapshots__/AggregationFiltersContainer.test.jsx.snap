// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregationFiltersContainer does not render aggregations with empty buckets 1`] = `
<div
  className="__AggregationFiltersContainer__ bg-white pa3"
>
  <div
    key="agg"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
          },
        ]
      }
      name="Jessica Jones"
      onChange={[Function]}
      range={false}
      selections={Array []}
    />
  </div>
</div>
`;

exports[`AggregationFiltersContainer renders initial state with initial url query q param 1`] = `
<div
  className="__AggregationFiltersContainer__ bg-white pa3"
>
  <div
    key="agg1"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
          },
        ]
      }
      name="Jessica Jones"
      onChange={[Function]}
      range={false}
      selections={
        Array [
          "agg1-selected",
        ]
      }
    />
  </div>
  <div
    key="agg2"
  >
    <AggregationFilter
      buckets={
        Immutable.List [
          Immutable.Map {
          },
        ]
      }
      name="Luke Cage"
      onChange={[Function]}
      range={false}
      selections={Array []}
    />
  </div>
</div>
`;
